### DISCUSSION TITLE
How to handle Generic Data

### DISCUSSION THREAD

**Discussion by AzHicham** [apalis-dev/apalis#256] [Q&A] [answered]
Hello

I'm facing a compile issue with this code, may be you have a solution ? :) 

I'm trying to create a generic worker registration.

```rust
fn register_worker<T: Model + 'static>() {
// Create my model
let model = T::Model::new();
// Create the worker
    let worker =
        WorkerBuilder::new("worker")
            .data(Arc::new(model)) // Add the model in shared state
            .with_storage(product_storage)
            .build_fn(handle_product_task::<T>); // Generic handler over the model

....
}
```

The issue only occures with the line  `.data(Arc::new(model)) ` when I remove it everything works pretty well. 

Here is the compile error 
```
error[E0599]: the method `build_fn` exists for struct `WorkerBuilder<_, RedisStorage<_>, Stack<Data<Arc<<T as ModelGenerator>::Model>>, Identity>, _>`, but its trait bounds were not satisfied
  --> src/bin/redis.rs:74:14
   |
69 | /         WorkerBuilder::new("worker")
70 | |             .data::<Arc<T::Model>>(Arc::new(model))
71 | |             .with_storage(product_storage)
72 | |             .build_fn(handle_product_task::<T>);
   | |             -^^^^^^^^ method cannot be called due to unsatisfied trait bounds
   | |_____________|
   | 
   |
   |
23 |   pub struct WorkerBuilder<Req, Source, Middleware, Serv> {
   |   -------------------------------------------------------
   |   |
   |   doesn't satisfy `_: WorkerFactory<_, ServiceFn<_, _>>`
   |   doesn't satisfy `_: WorkerFactoryFn<_, _, _>`
   |
   = note: the following trait bounds were not satisfied:
           `WorkerBuilder<_, RedisStorage<_>, Stack<apalis::prelude::Data<Arc<<T as ModelGenerator>::Model>>, Identity>, _>: WorkerFactory<_, apalis::prelude::ServiceFn<_, _>>`
           which is required by `WorkerBuilder<_, RedisStorage<_>, Stack<apalis::prelude::Data<Arc<<T as ModelGenerator>::Model>>, Identity>, _>: WorkerFactoryFn<_, _, _>`
           `&WorkerBuilder<_, RedisStorage<_>, Stack<apalis::prelude::Data<Arc<<T as ModelGenerator>::Model>>, Identity>, _>: WorkerFactory<_, apalis::prelude::ServiceFn<_, _>>`
           which is required by `&WorkerBuilder<_, RedisStorage<_>, Stack<apalis::prelude::Data<Arc<<T as ModelGenerator>::Model>>, Identity>, _>: WorkerFactoryFn<_, _, _>`
           `&mut WorkerBuilder<_, RedisStorage<_>, Stack<apalis::prelude::Data<Arc<<T as ModelGenerator>::Model>>, Identity>, _>: WorkerFactory<_, apalis::prelude::ServiceFn<_, _>>`
           which is required by `&mut WorkerBuilder<_, RedisStorage<_>, Stack<apalis::prelude::Data<Arc<<T as ModelGenerator>::Model>>, Identity>, _>: WorkerFactoryFn<_, _, _>`
```

I tried replacing model by a simple u16 it works.
It's like my object Model cannot be shared ... may be I have to constraint it more but with which trait ? 

Thanks a lot

**Comment by geofmureithi** [comment, marked as answer]
Tower offers a method to check if the service is actually clone.
https://docs.rs/tower/latest/tower/struct.ServiceBuilder.html#method.check_service_clone

From first look, I am curious if it's because you are missing `T: ?Sized` which is required by Arc to be clone.

Try these two and get back if they don't help.

**Comment by AzHicham** [comment]
Thanks for your help 

I still did not try [check_service_clone] (https://docs.rs/tower/latest/tower/struct.ServiceBuilder.html#method.check_service_clone), I need to figure out how to use it. 

I tried with ?Sized but it doesn't work. 

May be my traits are too constrained: 

Here is the full code :

```rust
pub trait ModelGenerator: Product + Sized {
    type Model: Model<Self, Err = Self::Err>;
    type Err: Error + Into<AppError>;

    fn model() -> Result<Self::Model, Self::Err> {
        Self::Model::new()
    }
}

pub trait Model<T: Product>: Sized {
    type Err: Error + Into<AppError>;

    fn new() -> Result<Self, Self::Err>;

    fn version(&self) -> Result<String, Self::Err>;
    ...
}

fn register_worker<T: ModelGenerator + 'static>() {
// Create my model
let model = T::model().unwrap();
// Create the worker
    let worker =
        WorkerBuilder::new("worker")
            .data(Arc::new(model)) // Add the model in shared state
            .with_storage(product_storage)
            .build_fn(handle_product_task::<T>); // Generic handler over the model

....
}
```

**Comment by AzHicham** [comment]
I also tried with less constrained traits (without Sized) but still got the same error: 

```rust
pub trait ModelGenerator: Product {
    type Model: Model;
    type Err: Error + Into<AppError>;

    fn model() -> Result<Self::Model, Self::Err>;
}

pub trait Model {
    type Product: Product;
    type Err: Error + Into<AppError>;

    fn version(&self) -> Result<String, Self::Err>;

    fn process(
        &self,
        input: (
            <Self::Product as Product>::ModelInput,
            <Self::Product as Product>::RenderInput,
        ),
        callback: ComposeCallback,
    ) -> Result<(<Self::Product as Product>::ModelOutput, String), Self::Err>;
}
```

In fact you can recreate this issue with a simple trait: 

```rust
pub trait BasicTrait {}

fn register_worker<T: BasicTrait>(data: T) {
    let worker =
        WorkerBuilder::new("worker")
            .data(Arc::new(data)) // Add the data in shared state
            .with_storage(product_storage)
            .build_fn(handle_product_task::<T>); // Generic handler over the model

}
```

It's like generic cannot be mixed with data and worker builder

**Comment by geofmureithi** [comment]
Can you try wrapping the Arc in a wrapper struct?
Something like `MyStruct(Arc)` and ensure you derive clone.

**Comment by geofmureithi** [comment]
To test check clone, use the `.chain` method to get the inner service builder. Ensure you call it after the data call.
It would be something like 
```rust
.chain(|svc| svc.check_svc(...))
```

**Comment by AzHicham** [comment]
Amazing it's working, using `check_service_clone` the compiler find out that Clone & Sync where missing.

I do understand why Sync is needed because Arc<T> is Sync only if T is Sync. 
What I do not understand yet is why T need to be cloned (in my case Model) ? 
IMO Arc need to be Clone but it's the [case](https://doc.rust-lang.org/std/sync/struct.Arc.html#impl-Clone-for-Arc%3CT,+A%3E). 

```rust
#[derive(Clone)]
MyStruct<T: Model> {
   model: Arc<T> // It should not be mandatory for T to impl Clone
}
```

**Comment by AzHicham** [comment]
nevermind I found out that using `#[derive(Clone)]` over `MyStruct` doesn't really work ... 
So I manually impl Clone for `MyStruct`

```rust
impl<T: ModelFactory> Clone for Ctx<T>
{
    fn clone(&self) -> Self {
        Self {
            model: self.model.clone(),
        }
    }
}
```
It's working perfectly :) 


Thank again for all your help !!!!

**Comment by geofmureithi** [comment]
Always happy to help. 🙂

**Comment by killix** [comment]
Hey @AzHicham, I have a bit of the same problem. I'm creating a system allowing Workers to be registered from a plugin system. And I'm a little stuck around Generic Data. Do you have any ideas for me? :) Thanks

**Comment by AzHicham** [comment]
> Hey @AzHicham, I have a bit of the same problem. I'm creating a system allowing Workers to be registered from a plugin system. And I'm a little stuck around Generic Data. Do you have any ideas for me? :) Thanks

Hello, sorry for the late reply 
I can try do you have some code to share ?
